### PR TITLE
LG-2527 Creates `StaticWidthText` component

### PR DESCRIPTION
**Pull request by TheSonOfThomp**
## ✍️ Proposed changes

Creates `StaticWidthText` component (within Typography package)

A wrapper component to ensure that updating the font weight does not affect the width of the element, and prevents layout shift should the font weight change. Useful for showing a bold effect on hover, focus, or active (See `Tabs`, `MenuItem`)

🎟 _Jira ticket:_ [PD-2527](https://jira.mongodb.org/browse/PD-2527)

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Style change (purely visual updates; if this is the case, attach a screenshot below!)

## ✅ Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README

## Feedback Levels

Please try to mark your comments with one of these levels

1. Praise (👏 🚀 🔥 🍾)
2. Opinions, Ideas, Suggestions (💡 👀)
3. Let’s discuss @ Code review (🤷 🤔 🧐 )
4. Doesn’t meet agreed upon standards (😬 ⚠️ ‼️)
5. Missing an edge case (🛑 🙅 ❌)
6. Missing core functionality (⛔️ ☣️ 🚧 🌋 💀)

(Note: Use ⌘-⌃-␣ to open the Emoji keyboard!)

### CODE REVIEW

File: packages/typography/src/utility/StaticWidthText.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+const staticWidthTextStyle = ({
+  pseudoElement,
+  maxFontWeight,
+}: Pick<StaticWidthTextProps<any>, 'pseudoElement' | 'maxFontWeight'>) => {
+  const pseudoSelector = `&:${pseudoElement}`;
+
+  return css`
+    /* Inherit relevant properties from the parent */
+    font-weight: inherit;
+    font-size: inherit;
```
Comment (shaneeza) [inline packages/typography/src/utility/StaticWidthText.tsx:29]: Not related to the component, but do you prefer splitting up your CSS based on properties?

File: packages/typography/src/Typography.story.tsx [hunk 2]
```diff
@@ -12,11 +12,14 @@ import {
   Label,
   Description,
   Link,
+  StaticWidthText,
```
Comment (shaneeza) [inline packages/typography/src/Typography.story.tsx:15]: I think if you read this name for the first time you wouldn't understand what it does. I'm not sure if this is any better or not but maybe something like, `InteractiveTextWidth`?

File: packages/typography/src/utility/StaticWidthText.tsx (hunk 1 above)
Comment (TheSonOfThomp) [reply packages/typography/src/utility/StaticWidthText.tsx:29]: Yeah, I try to follow (roughly) the categories in [idiomatic-css](https://github.com/necolas/idiomatic-css#declaration-order)

File: packages/typography/src/utility/StaticWidthText.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ */
+export function StaticWidthText<T extends keyof JSX.IntrinsicElements>({
+  as = 'span' as T,
+  children,
+  maxFontWeight = 700,
+  pseudoElement = 'after',
+  className,
+  ...rest
+}: StaticWidthTextProps<T>) {
+  const textValue = getNodeTextContent(children);
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply packages/typography/src/utility/StaticWidthText.tsx:84]: That's a good point, we should type children as `string` only

File: packages/typography/src/utility/StaticWidthText.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  const pseudoSelector = `&:${pseudoElement}`;
+
+  return css`
+    /* Inherit relevant properties from the parent */
+    font-weight: inherit;
+    font-size: inherit;
+
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
```
Comment (shaneeza) [inline packages/typography/src/utility/StaticWidthText.tsx:39]: I think you should also add in the comments that this component adds an ellipsis when text overflows

**Review by shaneeza — APPROVED** [review]
Exciting! I left another comment and looks like you're missing a changeset.

**Comment by shaneeza** [comment]
@TheSonOfThomp this one is missing a changeset

**Comment by TheSonOfThomp** [comment]
> this one is missing a changeset

Ah! Good catch!

**Pull request merged**